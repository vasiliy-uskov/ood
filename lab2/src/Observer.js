"use strict";
exports.__esModule = true;
var typescript_map_1 = require("typescript-map");
var Observable = /** @class */ (function () {
    function Observable() {
        this._observers = new typescript_map_1.TSMap();
        this._observersRemoveList = [];
    }
    Observable.prototype.addObserver = function (observer) {
        this._observers.set(observer, true);
    };
    Observable.prototype.removeObserver = function (observer) {
        this._observers.set(observer, false);
        this._observersRemoveList.push(observer);
    };
    Observable.prototype.notifyObservers = function () {
        var observers = this._observers.keys();
        for (var _i = 0, observers_1 = observers; _i < observers_1.length; _i++) {
            var observer = observers_1[_i];
            if (this._observers.get(observer)) {
                observer.update(this._getObservableData());
            }
        }
        ;
        this._clearRemoveList();
    };
    Observable.prototype._clearRemoveList = function () {
        for (var _i = 0, _a = this._observersRemoveList; _i < _a.length; _i++) {
            var observer = _a[_i];
            this._observers["delete"](observer);
        }
        this._observersRemoveList = [];
    };
    return Observable;
}());
exports.Observable = Observable;
