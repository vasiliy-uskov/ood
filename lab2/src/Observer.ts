import { TSMap } from "typescript-map"

interface IObserver<T> {
    update(data: T):void;
}

abstract class Observable<T> {
    public addObserver(observer: IObserver<T>, priority: number = this._greaterPriority) {
        this._greaterPriority = priority > this._greaterPriority ? priority : this._greaterPriority;
        this._observers.set(observer, {
            active: true,
            priority,
        });
    }

    public removeObserver(observer: IObserver<T>) {
        this._observers.get(observer).active = false;
        this._observersRemoveList.push(observer);
    }

    public notifyObservers() {
        const sortedObservers = this._observers.keys().slice();
        sortedObservers.sort((observer1: IObserver<T>, observer2: IObserver<T>) => {
            const priority1 = this._observers.get(observer1).priority;
            const priority2 = this._observers.get(observer2).priority;
            return priority1 > priority2 ? 1 : (priority1 == priority2 ? 0 : -1);
        });
        for (const observer of sortedObservers) {
            if (this._observers.get(observer).active) {
                observer.update(this._getObservableData());
            }
        }
        this._clearRemoveList();
    }

    protected abstract _getObservableData() : T

    private _clearRemoveList() {
        for (const observer of this._observersRemoveList) {
            this._observers.delete(observer);
        }
        this._observersRemoveList = [];
    }

    private _observers: TSMap<IObserver<T>, {active: boolean, priority: number}> = new TSMap();
    private _observersRemoveList: Array<IObserver<T>> = [];
    private _greaterPriority: number = 0;
}

export {IObserver};
export {Observable};