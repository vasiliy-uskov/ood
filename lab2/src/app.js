"use strict";
exports.__esModule = true;
var WhetherData_1 = require("./WhetherData");
var WhetherData_2 = require("./WhetherData");
var wd = new WhetherData_1.WhetherData();
var display = new WhetherData_2.StatisticsDisplay();
wd.addObserver(display);
wd.setMeasurements(3, 0.7, 760);
wd.setMeasurements(4, 0.8, 761);
wd.removeObserver(display);
wd.setMeasurements(10, 0.8, 761);
wd.setMeasurements(-10, 0.8, 761);
