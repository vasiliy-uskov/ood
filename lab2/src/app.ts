import {WhetherData} from "./WhetherData";
import {StatisticsDisplay} from "./WhetherData";

const wd: WhetherData = new WhetherData();
const display: StatisticsDisplay =  new StatisticsDisplay();
wd.addObserver(display);
wd.setMeasurements(3, 0.7, 760);
wd.setMeasurements(4, 0.8, 761);
wd.removeObserver(display);
wd.setMeasurements(10, 0.8, 761);
wd.setMeasurements(-10, 0.8, 761);