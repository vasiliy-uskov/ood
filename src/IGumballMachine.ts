export interface IGumballMachine {
	ejectQuarter(): void;
	insertQuarter(): void;
	turnCrank(): void;
	refill(count: number): void;
}