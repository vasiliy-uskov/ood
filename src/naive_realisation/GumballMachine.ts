import {State, stateToString} from "./Sates";
import {ILogger} from "../ILogger";

export class GumballMachine {
	private _state: State;
	private _count: number;
	private _logger: ILogger;

	constructor(count: number, logger: ILogger) {
		this._state = count > 0 ? State.NoQuarter : State.SoldOut;
		this._count = count;
		this._logger = logger;
	}

	insertQuarter(): void {
		switch (this._state) {
			case State.SoldOut:
				this._logger.log("You can't insert a quarter, the machine is sold out");
				break;
			case State.NoQuarter:
				this._logger.log("You inserted a quarter");
				this._state = State.HasQuarter;
				break;
			case State.HasQuarter:
				this._logger.log("You can't insert another quarter");
				break;
		}
	}

	ejectQuarter(): void {
		switch (this._state) {
			case State.HasQuarter:
				this._logger.log("Quarter returned");
				this._state = State.NoQuarter;
				break;
			case State.NoQuarter:
				this._logger.log("You haven't inserted a quarter");
				break;
			case State.SoldOut:
				this._logger.log("You can't eject, you haven't inserted a quarter yet");
				break;
		}
	}

	turnCrank(): void {
		switch (this._state) {
			case State.SoldOut:
				this._logger.log("You turned but there's no gumballs");
				break;
			case State.NoQuarter:
				this._logger.log("You turned but there's no quarter");
				break;
			case State.HasQuarter:
				this._logger.log("You turned...");
				this._logger.log("A gumball comes rolling out the slot...");
				--this._count;
				if (this._count == 0) {
					this._logger.log("Oops, out of gumballs");
					this._state = State.SoldOut;
				}
				else {
					this._state = State.NoQuarter;
				}
				break;
		}
	}

	refill(numBalls: number): void {
		this._count = numBalls;
		this._state = numBalls > 0 ? State.NoQuarter : State.SoldOut;
	}

	toString(): string {
		return `
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2019
Inventory: ${this._count} gumball${this._count != 1 ? "s" : ""}
Machine is ${stateToString(this._state)}
`;
	}
}