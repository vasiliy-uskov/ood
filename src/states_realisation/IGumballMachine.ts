export interface IGumballMachine {
	releaseBall(): void;

	getBallCount(): number;

	setSoldOutState(): void;

	setNoQuarterState(): void;

	setHasQuarterState(): void;
}
