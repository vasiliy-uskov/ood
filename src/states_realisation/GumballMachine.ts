import {IState} from "./IState";
import {IGumballMachine} from "./IGumballMachine";
import {SoldOutState} from "./SoldOutState";
import {NoQuarterState} from "./NoQuarterState";
import {HasQuarterState} from "./HasQuarterState";
import {ILogger} from "../ILogger";

export class GumballMachine {
	private readonly _logger: ILogger;
	private readonly _soldOutState: SoldOutState;
	private readonly _noQuarterState: NoQuarterState;
	private readonly _hasQuarterState: HasQuarterState;
	private _state: IState;
	private _count = 0;

	constructor(numBalls: number, logger: ILogger) {
		this._logger = logger;
		this._count = numBalls;
		const interfaceImpl = this._getGumballMachineInterface();
		this._soldOutState = new SoldOutState(interfaceImpl, logger);
		this._noQuarterState = new NoQuarterState(interfaceImpl, logger);
		this._hasQuarterState = new HasQuarterState(interfaceImpl, logger);
		if (this._count > 0) {
			this._state = this._noQuarterState;
		}
		else {
			this._state = this._soldOutState;
		}
	}

	ejectQuarter() {
		this._state.ejectQuarter();
	}

	insertQuarter() {
		this._state.insertQuarter();
	}

	turnCrank() {
		this._state.turnCrank();
	}

	refill(count: number) {
		this._count = count;
		if (count > 0) {
			this._setNoQuarterState();
		}
		else {
			this._setSoldOutState();
		}
	}

	toString(): string {
		return `
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2019
Inventory: ${this._count} gumball${this._count != 1 ? "s" : ""}
Machine is ${this._state.toString()}
`;
	}

	private _getGumballMachineInterface(): IGumballMachine {
		return {
			getBallCount: this._getBallCount.bind(this),
			releaseBall: this._releaseBall.bind(this),
			setSoldOutState: this._setSoldOutState.bind(this),
			setNoQuarterState: this._setNoQuarterState.bind(this),
			setHasQuarterState: this._setHasQuarterState.bind(this),
		}
	}

	private _getBallCount(): number {
		return this._count;
	}

	private _releaseBall(): void {
		if (this._count != 0) {
			this._logger.log("A gumball comes rolling out the slot...");
			--this._count;
		}
	}

	private _setSoldOutState(): void {
		this._state = this._soldOutState;
	}

	private _setNoQuarterState(): void {
		this._state = this._noQuarterState;
	}

	private _setHasQuarterState(): void {
		this._state = this._hasQuarterState;
	}
}