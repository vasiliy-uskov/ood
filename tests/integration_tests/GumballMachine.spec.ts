import {MockLogger} from "../MockLogger";
import {GumballMachine as StatesGumballMachine} from "../../src/states_realisation/GumballMachine";
import {GumballMachine as NaiveGumballMachine} from "../../src/naive_realisation/GumballMachine";
import {IGumballMachine} from "../../src/IGumballMachine";

function createStatesGumballMachine(startGumballsCount: number): {logger: MockLogger, machine: IGumballMachine} {
	const logger = new MockLogger();
	const machine = new StatesGumballMachine(startGumballsCount, logger);
	return {logger, machine}
}

function createNaiveGumballMachine(startGumballsCount: number): {logger: MockLogger, machine: IGumballMachine} {
	const logger = new MockLogger();
	const machine = new NaiveGumballMachine(startGumballsCount, logger);
	return {logger, machine}
}

function getMachines(startGumballsCount: number): Array<{logger: MockLogger, machine: IGumballMachine}> {
	return [
		createNaiveGumballMachine(startGumballsCount),
		createStatesGumballMachine(startGumballsCount),
	]
}

function checkAllElementsEquals<T>(items: Array<T>): void {
	if (!items.length) {
		return;
	}
	for (const item of items)
	{
		expect(item).toBe(items[0]);
	}
}

function useCase(message: string, startGumballsCount: number, useCaseFn: (machine: IGumballMachine) => void) {
	it(message, () => {
		const machines = getMachines(startGumballsCount);
		for (const {machine} of machines) {
			useCaseFn(machine);
		}
		const loggerResults = machines.map(({logger}) => logger.result());
		const states = machines.map(({machine}) => machine.toString());
		checkAllElementsEquals(loggerResults);
		checkAllElementsEquals(states);
		expect(states[0]).toMatchSnapshot();
		expect(loggerResults[0]).toMatchSnapshot();
	});
}

useCase('Если вставить четвертак и провернуть ручку, машина отдаёт жвачку', 1, machine => {
	machine.insertQuarter();
	machine.turnCrank();
});

useCase('После первого шарика жвачки можно получить следующий, если в автомате есть жвачка', 2, machine => {
	machine.insertQuarter();
	machine.turnCrank();
	machine.insertQuarter();
	machine.turnCrank();
});

useCase('Можно забрать четвертак, после того как он был вставлен', 1, machine => {
	machine.insertQuarter();
	machine.ejectQuarter();
});

useCase('Можно забрать несколько четвертаков, после того как они были вставлены', 1, machine => {
	machine.insertQuarter();
	machine.insertQuarter();
	machine.insertQuarter();
	machine.ejectQuarter();
});

useCase('Можно вставлять до пяти четвертаков', 1, machine => {
	machine.insertQuarter();
	machine.insertQuarter();
	machine.insertQuarter();
	machine.insertQuarter();
	machine.insertQuarter();
	machine.insertQuarter();
});

useCase('Если количество вставленных четвертаков больше, чем жвачки, хранящейся в автомате, то можно вернуть оставшиеся четвертаки', 1, machine => {
	machine.insertQuarter();
	machine.insertQuarter();
	machine.turnCrank();
	machine.turnCrank();
	machine.ejectQuarter();
	machine.turnCrank();
});

useCase('Можно заполнить автомат', 0, machine => {
	machine.refill(1);
});

useCase('Можно достать всю жвачку из автомата', 2, machine => {
	machine.refill(0);
});

useCase('При заполнении автомата, вставленные четвертаки не теряются', 1, machine => {
	machine.insertQuarter();
	machine.insertQuarter();
	machine.insertQuarter();
	machine.turnCrank();
	machine.turnCrank();
	machine.refill(5);
	machine.turnCrank();
	machine.turnCrank();
});


useCase('Если ручка была провёрнута четвертак изъять нельзя', 1, machine => {
	machine.insertQuarter();
	machine.turnCrank();
	machine.ejectQuarter();
});

useCase('Нельзя вставить два четвертака за раз', 2, machine => {
	machine.insertQuarter();
	machine.insertQuarter();
});

useCase('Нельзя достать четвертак, если он не был всавлен', 2, machine => {
	machine.ejectQuarter();
});


useCase('Нельзя провернуть ручку, если четвертак не был вставлен', 1, machine => {
	machine.turnCrank();
});

useCase('Нельзя получить жвачку если она закочилась', 0, machine => {
	machine.insertQuarter();
	machine.turnCrank();
});