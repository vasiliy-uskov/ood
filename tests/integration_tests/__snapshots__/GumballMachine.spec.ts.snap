// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Если вставить четвертак и провернуть ручку, машина отдаёт жвачку 1`] = `
"
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2019
Inventory: 0 gumballs
Machine is sold out
"
`;

exports[`Если вставить четвертак и провернуть ручку, машина отдаёт жвачку 2`] = `
"You inserted a quarter
You turned...
A gumball comes rolling out the slot...
Oops, out of gumballs
"
`;

exports[`Если ручка была провёрнута четвертак изъять нельзя 1`] = `
"
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2019
Inventory: 0 gumballs
Machine is sold out
"
`;

exports[`Если ручка была провёрнута четвертак изъять нельзя 2`] = `
"You inserted a quarter
You turned...
A gumball comes rolling out the slot...
Oops, out of gumballs
You can't eject, you haven't inserted a quarter yet
"
`;

exports[`Можно достать всю жвачку из автомата 1`] = `
"
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2019
Inventory: 0 gumballs
Machine is sold out
"
`;

exports[`Можно достать всю жвачку из автомата 2`] = `""`;

exports[`Можно забрать четвертак, после того как он был вставлен 1`] = `
"
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2019
Inventory: 1 gumball
Machine is waiting for quarter
"
`;

exports[`Можно забрать четвертак, после того как он был вставлен 2`] = `
"You inserted a quarter
Quarter returned
"
`;

exports[`Можно заполнить автомат 1`] = `
"
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2019
Inventory: 1 gumball
Machine is waiting for quarter
"
`;

exports[`Можно заполнить автомат 2`] = `""`;

exports[`Нельзя вставить два четвертака за раз 1`] = `
"
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2019
Inventory: 2 gumballs
Machine is waiting for turn of crank
"
`;

exports[`Нельзя вставить два четвертака за раз 2`] = `
"You inserted a quarter
You can't insert another quarter
"
`;

exports[`Нельзя достать четвертак, если он не был всавлен 1`] = `
"
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2019
Inventory: 2 gumballs
Machine is waiting for quarter
"
`;

exports[`Нельзя достать четвертак, если он не был всавлен 2`] = `
"You haven't inserted a quarter
"
`;

exports[`Нельзя получить жвачку если она закочилась 1`] = `
"
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2019
Inventory: 0 gumballs
Machine is sold out
"
`;

exports[`Нельзя получить жвачку если она закочилась 2`] = `
"You can't insert a quarter, the machine is sold out
You turned but there's no gumballs
"
`;

exports[`Нельзя провернуть ручку, если четвертак не был вставлен 1`] = `
"
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2019
Inventory: 1 gumball
Machine is waiting for quarter
"
`;

exports[`Нельзя провернуть ручку, если четвертак не был вставлен 2`] = `
"You turned but there's no quarter
"
`;

exports[`После первого шарика жвачки можно получить следующий, если в автомате есть жвачка 1`] = `
"
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2019
Inventory: 0 gumballs
Machine is sold out
"
`;

exports[`После первого шарика жвачки можно получить следующий, если в автомате есть жвачка 2`] = `
"You inserted a quarter
You turned...
A gumball comes rolling out the slot...
You inserted a quarter
You turned...
A gumball comes rolling out the slot...
Oops, out of gumballs
"
`;
